shell/imports/server/accounts/ldap.js

```diff
@@ -0,0 +1,240 @@
+      } else {
+        username = options.username;
+      }
+    }
+
+    if (_this.options.searchBindDn) {
+      let ldapBindFut = new Future();
+      client.bind(_this.options.searchBindDn, _this.options.searchBindPassword,
+        function (err) {
+          if (err) {
+            ldapBindFut.throw(err);
+          } else {
+            ldapBindFut.return();
+          }
+        }
+      );
+
+      try {
+        ldapBindFut.wait();
+      } catch (err) {
+        return { error: err };
+      }
+    }
+    // initialize result
+    let retObject = {
+      username: username,
+      email: domain ? username + "@" + domain : false,
+      emptySearch: true,
+      searchResults: {},
+    };
+
+    let filter = _this.options.filter;
+    Object.keys(_this.options.searchBeforeBind).forEach(function (searchKey) {
+      filter = "&" + filter + "(" + searchKey + "=" + _this.options.searchBeforeBind[searchKey] + ")";
+    });
+
+    let searchOptions = {
+      scope: "sub",
+      sizeLimit: 1,
+      filter: filter,
+    };
+
+    // perform LDAP search to determine DN
+    client.search(_this.options.base, searchOptions, function (err, res) {
+      if (err) {
+        if (resolved) return;
+        resolved = true;
+        ldapAsyncFut.return({
+          error: err,
+        });
+        return;
+      }
+
+      retObject.emptySearch = true;
+      res.on("searchEntry", function (entry) {
+        retObject.dn = entry.objectName;
+        retObject.username = retObject.dn;
+        retObject.emptySearch = false;
+
+        retObject.searchResults = _.omit(entry.object, "userPassword");
+
+        if (options.hasOwnProperty("searchUsername")) {
+          // This was only a search, return immediately
+          resolved = true;
+          ldapAsyncFut.return(retObject);
+          return;
+        }
+
+        // use the determined DN to bind
+        client.bind(entry.objectName, options.ldapPass, function (err) {
+          try {
+            if (err) {
+              throw new Meteor.Error(err.code, err.message);
+            }              else {
+              resolved = true;
+              ldapAsyncFut.return(retObject);
+            }
+          }
+          catch (e) {
+            if (resolved) return;
+            resolved = true;
+            ldapAsyncFut.return({
+              error: e,
+            });
+          }
+        });
+      });
+
+      res.on("error", errorFunc);
+
+      // If no dn is found, return as is.
+      res.on("end", function (result) {
+        if (retObject.dn === undefined) {
+          resolved = true;
+          ldapAsyncFut.return(retObject);
+        }
+      });
+    });
+
+    return ldapAsyncFut.wait();
+
+  } else {
+    throw new Meteor.Error(403, "Missing LDAP Auth Parameter");
+  }
+
+};
+
+LDAP.prototype.updateUserQuota = function (db, user) {
+  const fallback = user.cachedStorageQuota || 0;
```

Comment:
The fallback needs to be an object of {storage, grains, compute} to match the non-fallback return value, right?